src/functions.c

```diff
@@ -549,10 +550,64 @@ static void fcallCommandGeneric(client *c, int ro) {
         return;
     }
 
+    if ((fi->f_flags & FUNCTION_FLAG_NO_CLUSTER) && server.cluster_enabled) {
+        addReplyError(c, "Can not run function on cluster, 'no-cluster' flag is set.");
+        return;
+    }
+
+    if ((fi->f_flags & FUNCTION_FLAG_DENY_OOM) && server.script_oom) {
+        addReplyError(c, "-OOM deny-oom flag is set on the function, "
+                         "can not run it when used memory > 'maxmemory'");
+        return;
+    }
+
+    if (server.masterhost && server.repl_state != REPL_STATE_CONNECTED &&
+        server.repl_serve_stale_data == 0 && !(fi->f_flags & FUNCTION_FLAG_ALLOW_STALE))
+    {
+        addReplyError(c, "-MASTERDOWN Link with MASTER is down, "
+                         "replica-serve-stale-data is set to 'no' "
+                         "and 'allow-stale' flag is not set on the function.");
+        return;
+    }
+
+    if (fi->f_flags & FUNCTION_FLAG_WRITE) {
+        /* Function has the 'write' flag, make sure:
+         * 1. we are not a readonly replica
+         * 2. no disc error detected
+         * 3. command is not 'fcall_ro' */
+        if (server.masterhost && server.repl_slave_ro && c->id != CLIENT_ID_AOF
+            && !(c->flags & CLIENT_MASTER))
+        {
+            addReplyError(c, "Can not run a function with write flag on readonly replica");
+            return;
+        }
+
+        int deny_write_type = writeCommandsDeniedByDiskError();
+        if (deny_write_type != DISK_ERROR_TYPE_NONE && server.masterhost == NULL) {
+            if (deny_write_type == DISK_ERROR_TYPE_RDB)
+                addReplyError(c, "-MISCONF Redis is configured to save RDB snapshots, "
+                                 "but it is currently not able to persist on disk. "
+                                 "So its impossible to run functions that has 'write' flag on.");
+            else
+                addReplyErrorFormat(c, "-MISCONF Redis is configured to persist data to AOF, "
+                                       "but it is currently not able to persist on disk. "
+                                       "So its impossible to run functions that has 'write' flag on. "
+                                       "AOF error: %s", strerror(server.aof_last_write_errno));
+            return;
+        }
+
+        if (ro) {
+            addReplyError(c, "Can not execute a function with write flag using fcall_ro.");
+            return;
+        }
+    }
+
     scriptRunCtx run_ctx;
 
     scriptPrepareForRun(&run_ctx, fi->li->ei->c, c, fi->name);
-    if (ro) {
+    if (ro || !(fi->f_flags & FUNCTION_FLAG_WRITE)) {
+        /* On fcall_ro or on functions that do not have the 'write'
+         * flag, we will not allow write commands. */
         run_ctx.flags |= SCRIPT_READ_ONLY;
```

Comment:
In a similar way to this it might be good to pre-set `run_ctx.flags` to the desired OOM mode based on the given function flags provided in pr #10093 where we introduce `redis.set_oom_mode()` to handle how `redis.call()`s in the script/function will react to OOM conditions.

Reply:
i agree that `redis.set_oom_mode()` and the `deny-oom` function flag can share code.
i.e. on Functions you can define that behavior in flags, and for EVAL you have to call a method.
not sure if that's what you meant though.

Reply:
In #10093 we have a few different OOM runtime handling options. These options are all relevant only if the `deny-oom` flag is off. But if it is off we stil don't know which default handling to do:
1) Disable all writes (default).
2) Disable `CMD_DENYOOM` commands (and enable them when dirty).
3) Disable `CMD_DENYOOM` commands (and abort when dirty).
I think we might want flags to control the default behavior under OOM condition. But thinking about it again, I think it's better to stick with the current default and re-think this in the future.

Reply:
so maybe we can have all these flags exposed in function creation.
let's look into that after the function flags PR is merged.
unless you think exposing `deny-oom` is wrong, but since it's already exposed to modules, we may not want to change that.